src/flight.rs

```diff
@@ -193,3 +202,204 @@ impl SqlFlightClient {
+/// 2. If the consumer continues polling, automatically retry the entire query from the beginning
+/// 3. Apply exponential backoff between retries (1s, 2s, 4s, etc.)
+/// 4. Stop retrying after reaching `max_retries` attempts
+///
+/// ## Consumer Options
+///
+/// **Option 1: Continue polling for automatic retry**
+/// ```text
+/// Poll 1: Ok(batch1)
+/// Poll 2: Ok(batch2)
+/// Poll 3: Err(ConnectionReset) → Consumer continues polling
+/// Poll 4: Ok(batch1) → Query restarted from beginning
+/// Poll 5: Ok(batch2)
+/// Poll 6: Ok(batch3)
+/// ...
+/// ```
+///
+/// **Option 2: Stop on error**
+/// ```text
+/// Poll 1: Ok(batch1)
+/// Poll 2: Ok(batch2)
+/// Poll 3: Err(ConnectionReset) → Consumer stops polling
+/// ```
+///
+/// ## Important Notes
+/// - The query restarts from the beginning on retry - previously yielded batches will be re-yielded
+/// - Non-retryable errors are returned immediately without retry attempts
+/// - Only connection resets and specific gRPC errors trigger retries
+///
+pub struct RetryableQueryStream {
+    client: Arc<SqlFlightClient>,
+    sql: Arc<String>,
+    params: Arc<Option<RecordBatch>>,
+    state: StreamState,
+    max_retries: u32,
+    retry_count: u32,
+}
+
+impl RetryableQueryStream {
+    pub fn new(client: Arc<SqlFlightClient>, sql: &str, params: Option<RecordBatch>) -> Self {
+        Self {
+            max_retries: client.max_retries,
+            client,
+            sql: Arc::new(sql.to_string()),
+            params: Arc::new(params),
+            state: StreamState::Ready,
+            retry_count: 0,
+        }
+    }
+}
+
+impl Stream for RetryableQueryStream {
+    type Item = Result<RecordBatch, SpiceClientError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        match &mut self.state {
+            StreamState::Ready => {
+                let client = Arc::clone(&self.client);
+                let sql = Arc::clone(&self.sql);
+                let params = Arc::clone(&self.params);
+
+                let fut = Box::pin(async move {
+                    client
+                        .query_with_params(&sql, params.as_ref().clone())
+                        .await
+                });
+
+                self.state = StreamState::Executing(fut);
+                cx.waker().wake_by_ref();
+                Poll::Pending
+            }
+            StreamState::Executing(fut) => match fut.as_mut().poll(cx) {
+                Poll::Ready(Ok(stream)) => {
+                    self.state = StreamState::Streaming(Box::pin(stream));
+                    cx.waker().wake_by_ref();
+                    Poll::Pending
+                }
+                Poll::Ready(Err(error)) => {
+                    let is_retryable = is_connection_reset_generic_error(&error)
+                        && self.retry_count < self.max_retries;
+                    self.state = StreamState::Error {
```

Comment:
If this is a non-retryable error, yield the error directly and change the state to `Terminated`